src/azure-stac-cli/stac/common/__blob_service.py

```diff
@@ -0,0 +1,167 @@
+                    container_name: str,
+                    file_name: str,
+                    file_path: str=None ) -> None:
+    """ Uploads a blob to the storage account
+    :param conn_str: Connection String to the Storage Account hosting the Blob
+    :type conn_str: str 
+    :param container_name: Name of the Container where the blob is hosted in the Azure Storage Account
+    :type container_name: str
+    :param file_path: Relative path to the blob with ref. to the container
+    :type file_path: str
+    :param file_name: Name of the blob in storage account
+    :type file_name: str
+    :returns: None
+    :rtype: None
+    """
+    blob = BlobClient.from_connection_string(
+        conn_str=conn_str, 
+        container_name=container_name, 
+        blob_name=f"{file_path}/{file_name}" if file_path != None else file_name)
+    
+    async with blob:
+        # [START upload_blob_to_container]
+        with open(file_name, "rb") as blob_data:
+            await blob.upload_blob(data=blob_data)
+
+# Data is being downloaded locally and will need to be cleaned up
+# by the calling module.
+async def download_blob_async(conn_str: str, 
+                              container_name: str, 
+                              file_path: str,
+                              destination_path: str) -> str:
+    
+    """ Download blob from azure storage account. Will not overwrite
+    if the file already exists
+    :param conn_str: Connection String to the Storage Account hosting the Blob
+    :type conn_str: str 
+    :param container_name: Name of the Container where the blob is hosted in the Azure Storage Account
+    :type container_name: str
+    :param file_path: Relative path to the blob with ref. to the container
+    :type file_path: str
+    :param destination_path: Absolute location to the local folder where the blob needs to be downloaded to
+    :type destination_path: str
+    :returns: Path to the local downloaded file
+    :rtype: str
+    """
+    
+    blob_name = os.path.basename(file_path)
+    
+    try:
+        async with BlobClient.from_connection_string(
+                conn_str=conn_str, 
+                container_name=container_name, 
+                blob_name=blob_name) as blob_client:
+
+            _, file_name = os.path.split(
+                blob_client.blob_name)
+
+            # get the full path for the destination where the file will be 
+            # downloaded including the leaf file name
+            download_file_path = os.path.join(
+                destination_path, file_name)
+            
+            does_blob_exist = await blob_client.exists()
+            
+            # check for blob existence
+            if does_blob_exist:
+                
+                with open(download_file_path, "wb") as fh:
+                    
+                    stream = await blob_client.download_blob()
+                    data = await stream.readall()
+                    fh.write(data)
```

Comment:
Won't this read the entire blob into memory and then write it to a local file? It would be better to stream it to disk in chunks (e.g., using the readinto method). That way we use a fixed small amount of memory.

Reply:
This is certainly a good place to optimize the processing. But for this PR, I have focused on restructuring and refactoring the code.
Added a task in this Sprint to optimize this logic. I will address this optimization in the upcoming PRs.